File: project_euler/Problem072/Problem072.java
package project_euler;

import java.util.HashMap;
import java.util.Map;

public class Problem072 {
    public void solution() {
        int target = 1000000;
        Map<Integer, Long> fractions = new HashMap<>();
        fractions.put(1, getSummation(1));
        long result = getFractions(target, fractions) - fractions.get(1);
        System.out.println(result);
    }

    private long getFractions(int target, Map<Integer, Long> fractions) {
        if (fractions.containsKey(target)) {
            return fractions.get(target);
        }

        int index = 2;
        long result = 0L;
        while (index <= target) {
            int step = target / (target / index);
            result += (step - index + 1) * getFractions(target / step, fractions);
            index = step + 1;
        }
        result = getSummation(target) - result;
        fractions.put(target, result);
        return result;
    }

    private long getSummation(int value) {
        return (long) value * (value + 1) >> 1;
    }
}

File: project_euler/Problem072/Problem072.py
class Problem072:
    def solution(self) -> None:
        target, fractions = 1000000, {1: self.__get_summation(1)}
        result = self.__get_fractions(target, fractions) - fractions[1]
        print(result)

    def __get_fractions(self, target: int, fractions: dict[int, int]) -> int:
        if target in fractions:
            return fractions[target]

        index, result = 2, 0
        while index <= target:
            step = target // (target // index)
            result += (step - index + 1) * self.__get_fractions(target // step, fractions)
            index = step + 1
        result = self.__get_summation(target) - result
        fractions[target] = result
        return result

    def __get_summation(self, value: int) -> int:
        return value * (value + 1) >> 1
